src/Enzo/enzo_EnzoMethodGrackle.cpp

```diff
@@ -512,17 +512,30 @@ double EnzoMethodGrackle::timestep ( Block * block ) throw()
       delete_cooling_time = true;
     }
 
+    // the grackle_units_ member can't be used here since this method MUST be
+    // const-qualified (i.e. the values of member variables can't be changed)
+    code_units temp_grackle_units;
     grackle_field_data grackle_fields_;
 
-    setup_grackle_units(enzo_block,  &grackle_units_);
+    setup_grackle_units(enzo_block,  &temp_grackle_units);
     setup_grackle_fields(enzo_block, &grackle_fields_);
 
-    if (calculate_cooling_time(&grackle_units_, &grackle_fields_, cooling_time) == ENZO_FAIL) {
+    if (calculate_cooling_time(&temp_grackle_units, &grackle_fields_, cooling_time) == ENZO_FAIL) {
       ERROR("EnzoMethodGrackle::compute()",
       "Error in calculate_cooling_time.\n");
     }
 
-    for (int i = 0; i < size; i++) dt = std::min(enzo_float(dt), std::abs(cooling_time[i]));
+    // make sure to exclude the ghost zone. Because there is no refresh before
+    // this method is called (at least during the very first cycle) - this can
+    // including ghost zones can lead to timesteps of 0
+    for (int iz = gz; iz < ngz - gz; iz++) {   // if rank < 3: gz = 0, ngz = 1
+      for (int iy = gy; iy < ngy - gy; iy++) { // if rank < 2: gy = 0, ngy = 1
+        for (int ix = gx; ix < ngx - gx; ix++) {
+	  int i = INDEX(ix, iy, iz, ngx, ngy);
```

Comment:
I'm not sure, but would it be better for the declaration of `i` to be outside the loop?

Reply:
I suspect the compiler is going to take care of any optimisation issue here and keeping the variable scope within the nest is good practice.

Reply:
I very much share John's sentiments. I feel like it conveys more information about the use of `i`. I doubt it has any performance implications (especially since this is an int and not something that requires heap allocation - like a string).  I think this probably just boils down to a question of coding style.

Reply:
Ok, fair enough. It's fine with me.

Reply:
I've done a little reading on this and it does seem that this is a better way to do things. There are apparently additional optimizations the compiler can do when things are done this. Thanks!